### PR TITLE
Add options to configure IPFS gateway to use to fetch data

### PR DESCRIPTION
**Pull request by mderriey**
### Description of change

This adds options to `CacheOnlyIPFS`, `PinataStorageWithCache`, and `PinataStorage` to specify which IPFS gateway to use when fetching data.

Options include the base URL of the gateway (assuming the content can always be fetched at `/ipfs/<cid>`).

### Pull-Request Checklist

- [X] Code is up-to-date with the `main` branch
- [X] `npm run lint` passes with this change
- [X] `npm test` passes with this change
- [X] `npm run commitlint` passes with this change
- [ ] This pull request links relevant issues as `Fixes #0000`
- [X] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change

**_Please note that this repository uses [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) in combination with [semantic-release](https://github.com/semantic-release/semantic-release) to automate package publication. Therefore, your commit messages are critical, and the build process will lint them. If the build fails at the commitlint stage, please refer to this [handy guide](https://www.git-tower.com/learn/git/faq/edit-fix-commit-message) on how to update the commit messages_**

### CODE REVIEW

File: src/ipfs.ts [hunk 1]
```diff
@@ -68,21 +68,21 @@ export class InMemoryIPFS implements IPFS {
 
 export class CacheOnlyIPFS implements IPFS {
   private cache: ObjectCache
-  private getFromIpfs: boolean
+  private ipfsGatewayOptions?: IPFSGatewayOptions
 
-  constructor(cache: ObjectCache, getFromIpfs?: boolean) {
+  constructor(cache: ObjectCache, ipfsGatewayOptions?: IPFSGatewayOptions) {
```
Comment (neilcampbell) [inline src/ipfs.ts:73]: This is a breaking change, so we should release as such.

File: src/ipfs.ts [hunk 2]
```diff
@@ -330,10 +340,18 @@ class NoOpCache implements ObjectCache {
...
-  constructor(pinataToken: string) {
-    super(pinataToken, new NoOpCache())
+  constructor(pinataToken: string, ipfsGatewayOptions?: IPFSGatewayOptions) {
+    super(pinataToken, new NoOpCache(), ipfsGatewayOptions)
   }
 }
+
+export function getCIDUrl(ipfsGatewayBaseUrl: URL, cid: string) {
+  const copy = new URL(ipfsGatewayBaseUrl.toString())
+  copy.pathname = `/ipfs/${cid}`
```
Comment (neilcampbell) [inline src/ipfs.ts:354]: Not all gateways use this format, for example cloudflare used a subdomain.

I think it's probably ok to assume it's path based (not what cloudflare does), however I'd be inclined to require the full url minus the cid, so in the default value that'd be `https://ipfs.algonode.dev/ipfs`

File: src/ipfs.ts (hunk 1 above)
Comment (mderriey) [reply src/ipfs.ts:73]: Agreed.

I'm not well-versed with the tooling around that, but saw after a quick look that the commit title `feat!: <bla>` (note the bang) should be recognised as one and trigger a major version bump.

Comment (neilcampbell) [reply src/ipfs.ts:73]: Unfortunately `feat!:` doesn't actually trigger a breaking change bump in the npm semantic release plugin.
You need to add a commit footer with the text "Breaking Change:", it's a pain!

Here's an example commit https://github.com/algorandfoundation/algokit-client-generator-ts/commit/dd9d15769f7aed39a1ed8d3df47cec6c0dd22beb

You may find it easy to squash the GitHub UI and add the breaking change info.

**Review by neilcampbell — APPROVED** [review]

File: src/ipfs.ts (hunk 2 above)
Comment (robdmoore) [reply src/ipfs.ts]: Cloudflare actually still worked with this format

**Pull request merged**